### PR TITLE
RqHref.Base validates Request-Line format

### PR DESCRIPTION
**Pull request by Happy-Neko**
Issue #445 fix and todo suggestion.
@davvd Please proceed with code review.

### CODE REVIEW

**Comment by davvd** [comment]
@happy-neko Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@dmzaytsev review this one,please

File: src/main/java/org/takes/rq/RqHref.java [hunk 1]
```diff
@@ -67,16 +79,36 @@
         public Base(final Request req) {
             super(req);
         }
+
+        // @todo #445 RqMethod already validates Request-Line and
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqHref.java]: @Happy-Neko the puzzle format should be like `#445:30min/DEV`
otherwise rultor won't merge the PR

**Comment by dmzaytsev** [comment]
@Happy-Neko just one comment above

**Comment by Happy-Neko** [comment]
@dmzaytsev todo updated

**Comment by dmzaytsev** [comment]
@Happy-Neko thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by Happy-Neko** [comment]
@yegor256 ping

File: src/main/java/org/takes/rq/RqHref.java [hunk 2]
```diff
@@ -67,16 +79,36 @@
...
         @Override
         public Href href() throws IOException {
+            if (!this.head().iterator().hasNext()) {
+                throw new HttpException(
+                    HttpURLConnection.HTTP_BAD_REQUEST,
+                    "HTTP Request should have Request-Line"
+                );
+            }
+            final String requestLine = this.head().iterator().next();
+            final Matcher uriMatcher =
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqHref.java]: @Happy-Neko why not just `matcher`? see http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

File: src/main/java/org/takes/rq/RqHref.java [hunk 3]
```diff
@@ -67,16 +79,36 @@
...
+        // in one place to enforce DRY principle.
         @Override
         public Href href() throws IOException {
+            if (!this.head().iterator().hasNext()) {
+                throw new HttpException(
+                    HttpURLConnection.HTTP_BAD_REQUEST,
+                    "HTTP Request should have Request-Line"
+                );
+            }
+            final String requestLine = this.head().iterator().next();
```
Comment (yegor256) [inline src/main/java/org/takes/rq/RqHref.java]: @Happy-Neko just `line` is not enough?

**Comment by yegor256** [comment]
@Happy-Neko two comments above

**Comment by Happy-Neko** [comment]
@yegor256 variables renamed.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by Happy-Neko** [comment]
@yegor256 I have added extra space before puzzle description. Please verify, I am not sure what the proper format is.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@ypshenychka please, review this ticket for compliance with our [QA rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy now pls

**Comment by ypshenychka** [comment]
@Happy-Neko 
>Issue #445 fix and todo suggestion.
@davvd Please proceed with code review.

According to our [QA Rules](http://www.teamed.io/qa.html):
>Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

Please correct your description by providing more information regarding what was done in the branch.

**Comment by ypshenychka** [comment]
@dmzaytsev 
According to our [QA Rules](http://www.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one cosmetic issue was found during code review.
Please confirm that you'll try to find at least three major problems while future reviews along with cosmetic concerns.

**Comment by Happy-Neko** [comment]
@ypshenychka description updated.

**Comment by ypshenychka** [comment]
@Happy-Neko Thank you.

**Comment by dmzaytsev** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@dmzaytsev thanks

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@dmzaytsev added 10 mins to @ypshenychka, for QA review, payment code is `72549980`... done, I added **20 mins** in payment `72549996`, 72 hours and 3 mins spent to complete this... extra minutes for review comments (c=5)... +20 to your rating, your total score is [+1038](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**